### ISSUE TITLE
wrong match when pattern uses equivalence class and 'breakindentopt' set to `list:-1`

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#9309]
**Steps to reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        &breakindent = true
        &breakindentopt = 'list:-1'
        &formatlistpat = '^-'
        &columns = 80
        '- y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'->setline(1)
        &hlsearch = true
        @/ = '\%>3v[[=y=]]'
        echomsg searchcount().total
    EOF
    )

Almost all the line in the buffer is matched by the pattern, and is highlighted with `Search`.  20 is echo'ed.

That's wrong.  The buffer only contains 2 `y` characters.  So – at most – the pattern can only have 2 matches.  And because of the `\%3v` atom, only the second `y` can match.

**Expected behavior**

Only the second `y` character is matched by the pattern, and is highlighted with `Search`.  1 is echo'ed.

**Operating system**

Ubuntu 20.04.3 LTS

**Version of Vim**

8.2 Included patches: 1-3762

**Additional Context**

The issue disappears if we remove the `\%v` atom from the pattern:

    # NOT ok
    @/ = '\%>3v[[=y=]]'

    # OK
    @/ = '[[=y=]]'

Test:
```vim
vim9script
&breakindent = true
&breakindentopt = 'list:-1'
&formatlistpat = '^-'
&columns = 80
'- y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'->setline(1)
&hlsearch = true
@/ = '[[=y=]]'
echomsg searchcount().total
```
    2

---

The issue disappears if we remove the equivalence class from the pattern:

    # NOT ok
    @/ = '\%>3v[[=y=]]'

    # OK
    @/ = '\%>3vy'

Test:
```vim
vim9script
&breakindent = true
&breakindentopt = 'list:-1'
&formatlistpat = '^-'
&columns = 80
'- y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'->setline(1)
&hlsearch = true
@/ = '\%>3vy'
echomsg searchcount().total
```
    1

---

The issue disappears if we use `\%c` instead of `\%v`:

    # NOT ok
    @/ = '\%>3v[[=y=]]'

    # OK
    @/ = '\%>3c[[=y=]]'

Test:
```vim
vim9script
&breakindent = true
&breakindentopt = 'list:-1'
&formatlistpat = '^-'
&columns = 80
'- y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'->setline(1)
&hlsearch = true
@/ = '\%>3c[[=y=]]'
echomsg searchcount().total
```
    1

---

The issue disappears if we remove `list:-1` from `'breakindentopt'`:
```vim
vim9script
&breakindent = true
&formatlistpat = '^-'
&columns = 80
'- y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx y xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'->setline(1)
&hlsearch = true
@/ = '\%>3v[[=y=]]'
echomsg searchcount().total
```
    1

**Comment by lacygoill** [comment]
This issue affects a user if they include `list:-1` in `'breakindentopt'` and use the [vim-sneak](https://github.com/justinmk/vim-sneak) plugin.  Because [the latter uses equivalence classes](https://github.com/justinmk/vim-sneak/blob/94c2de47ab301d476a2baec9ffda07367046bec9/autoload/sneak/search.vim#L11).

**Comment by lacygoill** [comment]
Here are some logs:

- [nfa_regexp_debug.log](https://github.com/vim/vim/files/7684919/nfa_regexp_debug.log)
- [nfa_regexp_dump.log](https://github.com/vim/vim/files/7684920/nfa_regexp_dump.log)
- [nfa_regexp_run.log](https://github.com/vim/vim/files/7684921/nfa_regexp_run.log)

Vim didn't write any `nfa_regexp_error.log` file.  I guess it means that Vim considers that no errors were encountered while computing the matches.

---

At some point, we might want to document how to generate log files for the regex engine:

    sed -i '\@^//\s*#define DEBUG$@ s@//@@' src/regexp.c

I'll try to send a patch later.

Also, I think there is a typo in `src/regexp_nfa.c`:
```diff
diff --git a/src/regexp_nfa.c b/src/regexp_nfa.c
index ffe894d4a..9dc246d28 100644
--- a/src/regexp_nfa.c
+++ b/src/regexp_nfa.c
@@ -18,7 +18,7 @@
  * The last three are enabled when compiled as debug mode and individually
  * disabled by commenting them out.
  * The log files can get quite big!
- * Do disable all of this when compiling Vim for debugging, undefine DEBUG in
+ * To disable all of this when compiling Vim for debugging, undefine DEBUG in
  * regexp.c
  */
 #ifdef DEBUG
```

**Comment by brammool** [comment]
Perhaps @chrisbra knows more about this?

**Comment by chrisbra** [comment]
Hm, 
it's hard to understand what is happening. It only seems to happen when some kind of a collation is being used, so even `\%>3v[y]` it happens but it doesn't happen with `\%>3vy`. 

Also, it's very hard to figure out what's happening, because the search highlighting code compiles its own regex when redrawing and at the same time, while redrawing to figure out the the correct indenting, we are compiling another regex for the 'formatlistpat'. I wonder if we need to save and restore some states when calling `get_breakindent_win()`, not sure.

Also, I noticed, the breakindenting for the list indenting is very in-efficient, because it does not make use of the cacheing, so it will compile the same pattern over and over again, as often as the `get_breakindent_win()` function is called. 

I can fix this (and PR coming for that, that will also fix a couple of minor problems I noticed) and this makes the obvious problem go away (at worst after the next redraw), but I don't understand yet why this only happens for collations and the NFA engine....
